example_error_handler_test.go

```diff
@@ -0,0 +1,113 @@
+package river_test
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+
+	"github.com/riverqueue/river"
+	"github.com/riverqueue/river/internal/riverinternaltest"
+	"github.com/riverqueue/river/internal/util/slogutil"
+	"github.com/riverqueue/river/riverdriver/riverpgxv5"
+)
+
+type CustomErrorHandler struct{}
+
+func (*CustomErrorHandler) HandleError(ctx context.Context, job *river.JobRow, err error) *river.ErrorHandlerResult {
+	fmt.Printf("Job errored with: %s\n", err)
+
+	// Either function can also set the job to be immediately cancelled, which
+	// we take advantage of here to make sure it's not retried in the example.
+	// Can also be `return nil`.
+	return &river.ErrorHandlerResult{SetCancelled: true}
+}
+
+func (*CustomErrorHandler) HandlePanic(ctx context.Context, job *river.JobRow, panicVal any) *river.ErrorHandlerResult {
+	fmt.Printf("Job panicked with: %v\n", panicVal)
+	return &river.ErrorHandlerResult{SetCancelled: true}
+}
+
+type ErroringArgs struct {
+	ShouldError bool
+	ShouldPanic bool
+}
+
+func (args ErroringArgs) Kind() string { return "erroring" }
+
+type ErroringWorker struct {
+	river.WorkerDefaults[ErroringArgs]
+}
+
+func (w *ErroringWorker) Work(ctx context.Context, j *river.Job[ErroringArgs]) error {
+	switch {
+	case j.Args.ShouldError:
+		return fmt.Errorf("this job errored")
+	case j.Args.ShouldPanic:
+		panic("this job panicked")
+	}
+	return nil
+}
+
+// Example_errorHandler demonstrates how to use the ErrorHandler interface for
+// custom application telemetry.
+func Example_errorHandler() {
+	ctx := context.Background()
+
+	dbPool, err := pgxpool.NewWithConfig(ctx, riverinternaltest.DatabaseConfig("river_testdb_example"))
+	if err != nil {
+		panic(err)
+	}
+	defer dbPool.Close()
+
+	// Required for the purpose of this test, but not necessary in real usage.
+	if err := riverinternaltest.TruncateRiverTables(ctx, dbPool); err != nil {
+		panic(err)
+	}
+
+	workers := river.NewWorkers()
+	river.AddWorker(workers, &ErroringWorker{})
+
+	riverClient, err := river.NewClient(riverpgxv5.New(dbPool), &river.Config{
+		ErrorHandler: &CustomErrorHandler{},
+		Logger:       slog.New(&slogutil.SlogMessageOnlyHandler{Level: 9}), // Suppress logging so example output is cleaner (9 > slog.LevelError).
```

Comment:
This is a bit unconventional, but there's a fair bit of error output on error/panic from `jobExecutor`. Setting level to > `slog.LevelError` suppresses it and makes the test output much cleaner, but in case of something bad happening might also suppress other things I suppose. Not sure which is the lesser of two evils, but this is the one I went with.

Reply:
The verbose error/panic output might be something we need to take a look at. I don't think there was much intentional design there, more that it just evolved that way as we pieced together functionality.

Reply:
True. I don't think we're _too_ far off though. If I was running this thing I'd expect to that there was reasonable error logging by default so problems aren't black holed without me having to chase down more configuration.